src/ui/marker.js

```diff
@@ -11,26 +11,54 @@ import type Popup from './popup';
 import type {LngLatLike} from "../geo/lng_lat";
 import type {MapMouseEvent} from './events';
 
+export type Anchor = 'top' | 'bottom' | 'left' | 'right' | 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right';
+export type Offset = number | PointLike | {[Anchor]: PointLike};
+
+export type MarkerOptions = {
+    anchor: Anchor,
+    offset: Offset
+}
+
 /**
  * Creates a marker component
- * @param element DOM element to use as a marker. If left unspecified a default SVG will be created as the DOM element to use.
- * @param options
- * @param options.offset The offset in pixels as a {@link PointLike} object to apply relative to the element's center. Negatives indicate left and up.
+ * @param {Object} [element] DOM element to use as a marker. If left unspecified a default SVG will be created as the DOM element to use.
+ * @param {Object} [options]
+ * @param {string} [options.anchor] - A string indicating the markers's location relative to
+ *   the coordinate set via {@link Marker#setLngLat}.
+ *   Options are `'top'`, `'bottom'`, `'left'`, `'right'`, `'top-left'`,
+ *   `'top-right'`, `'bottom-left'`, and `'bottom-right'`. If unset the anchor will be
+ *   dynamically set to ensure the marker falls within the map container with a preference
+ *   for `'bottom'`.
+ * @param {number|PointLike|Object} [options.offset] The offset in pixels as a {@link PointLike} object to apply relative to the element's center. Negatives indicate left and up.
  * @example
+ * var markerHeight = 50, markerRadius = 10, linearOffset = 25;
+ * var markerOffsets = {
```

Comment:
Popup has this since the anchor of the popup changes to ensure it's visible within the map (ie. moves below the point when there is no space above it), the Popup offset needs to be different for each anchor in some use cases.

Is there a use case for having a variable anchor for a single marker and hence needing different offsets set for each anchor?

Reply:
mmm I was thinking about exposing as a variable so users could define their own offset per marker/anchor. This could be for styling considerations on static maps or maps with markers in close proximity to eachother. 

One possible route to take would have popup inherit the anchor from marker as a default for when users don't define?

Reply:
Good point, I guess you could have different marker images anchored at different locations.

Reply:
I see, for example the default Marker in Mapbox with an anchor of bottom still needs an offset due to the shadow, so if you have a marker that goes upsidedown when it can't fit at the top of the map and it has a shadow it needs an offset per anchor value. Got it.

Reply:
yup! I think its good to have built in flexibility but include batteries for desired behaviors if not specified. These were great questions though, thank you for taking the time to review my work so far ✨🎷🐛

Reply:
As @andrewharvey noted above, the reason we need this in `Popup` is that it dynamically decides the anchor based on the popup's position relative to the edges of the viewport.  I don't think `Marker` should include such logic -- it's lower-level than `Popup`, and the use case for such automatic anchoring magic seems less clear.

Reply:
sorry to keep things held up!! i've been working my way through testing and will make another commit soon but would definitely welcome another pair of eyes on this n_n